**<h3> 1. IP ASSIGNMENTS <H3>**

In [ ]:
#NB:
#some cells have the sketchy(analysis) part of the solutions but for the final solutions, 
#i have labelled/commented in each cell(as, e.g #question 1)

**<h4> a) First connect to the Database <h4>**

In [ ]:
#load the sql extension into my working environment

%load_ext sql

# Then connect to the in memory sqlite database
%sql sqlite://

'Connected: @None'

**<h4>b) Importing Data From a CSV File<h4>**

In [ ]:
#Importing the necessary python libraries needed
#
import csv
import pandas as pd


#Loading the table from a local CSV file
with open('uganda.csv','r') as f:
    uganda = pd.read_csv(f, index_col=0, encoding='utf-8') 

#Then storing it in an SQL table of the memory sqlite database
%sql DROP TABLE if EXISTS uganda;
%sql PERSIST uganda;       

 * sqlite://
Done.
 * sqlite://


'Persisted uganda'

**<h4>c) Preview of the uganda table <h4>**

In [ ]:
# select all the columns in the table and limit to only the first five records
%sql SELECT * FROM uganda LIMIT 5;

 * sqlite://
Done.


[(263, 'KACHERI', 'KOTIDO', 17244, 1067176155, 'Y', 354.20741069999997, 1137.467019, 7023.533691, 6434.342449, 528.1242294, 2279091.779, 600723.8929),
 (264, 'KOTIDO', 'KOTIDO', 52771, 597575188, 'Y', 367.89052319999996, 1162.9966869999998, 13587.990759999999, 12455.59264, 824.7670809, 4582294.492, 959201.3825),
 (265, 'KOTIDO TOWN COUNCIL', 'KOTIDO', 27389, 23972401, 'Y', 369.3141771, 1167.005832, 1656.5318550000002, 1520.322052, 8.561643817, 561476.4874, 9991.488268000001),
 (266, 'NAKAPERIMORU', 'KOTIDO', 38775, 419111591, 'Y', 283.3245689, 852.3665784, 7087.823334000001, 6761.488901000001, 45.72171184, 1915695.928, 38971.65908),
 (267, 'PANYANGARA', 'KOTIDO', 65704, 880955930, 'Y', 373.8369255, 1283.859882, 10398.24939, 10111.198129999999, 172.6119139, 3779939.224, 221609.5114)]

           selecting the columns to better understand the data

In [ ]:
#select only the population column
%sql SELECT POP FROM uganda LIMIT 4;

 * sqlite://
Done.


[(17244,), (52771,), (27389,), (38775,)]

In [ ]:
#total number of subcounties
%%sql
SELECT COUNT(SUBCOUNTY_NAME)from uganda;

 * sqlite://
Done.


[(52,)]

In [ ]:
#select the area column only and limit it to four rows
%sql SELECT Area FROM uganda LIMIT 4;

 * sqlite://
Done.


[(1067176155,), (597575188,), (23972401,), (419111591,)]

In [ ]:
#select the karamoja column only and limit the records to 4
%sql SELECT Karamoja FROM uganda LIMIT 4;

 * sqlite://
Done.


[('Y',), ('Y',), ('Y',), ('Y',)]

In [ ]:
#select only the column for S_Prod_Tot, 4 records
%sql SELECT S_Prod_Tot FROM uganda LIMIT 4;

 * sqlite://
Done.


[(2279091.779,), (4582294.492,), (561476.4874,), (1915695.928,)]

In [ ]:
%%sql
SELECT SUBCOUNTY_NAME, DISTRICT_NAME
FROM Uganda;

 * sqlite://
Done.


[('KACHERI', 'KOTIDO'),
 ('KOTIDO', 'KOTIDO'),
 ('KOTIDO TOWN COUNCIL', 'KOTIDO'),
 ('NAKAPERIMORU', 'KOTIDO'),
 ('PANYANGARA', 'KOTIDO'),
 ('RENGEN', 'KOTIDO'),
 ('KAABONG  EAST', 'KAABONG'),
 ('KAABONG TOWN COUNCIL', 'KAABONG'),
 ('KAABONG WEST', 'KAABONG'),
 ('KALAPATA', 'KAABONG'),
 ('KAMION', 'KAABONG'),
 ('KAPEDO', 'KAABONG'),
 ('KATHILE', 'KAABONG'),
 ('KAWALAKOL', 'KAABONG'),
 ('LOBALANGIT', 'KAABONG'),
 ('LODIKO', 'KAABONG'),
 ('LOLELIA', 'KAABONG'),
 ('LOYORO', 'KAABONG'),
 ('NAPORE (KARENGA)', 'KAABONG'),
 ('SIDOK', 'KAABONG'),
 ('ABIM', 'ABIM'),
 ('ABIM TOWN COUNCIL', 'ABIM'),
 ('ALEREK', 'ABIM'),
 ('LOTUKEI', 'ABIM'),
 ('MORULEM', 'ABIM'),
 ('KARITA', 'AMUDAT'),
 ('LOROO', 'AMUDAT'),
 ('AMUDAT', 'AMUDAT'),
 ('AMUDAT TOWN COUNCIL', 'AMUDAT'),
 ('KATIKEKILE', 'MOROTO'),
 ('NADUNGET', 'MOROTO'),
 ('TAPAC', 'MOROTO'),
 ('NORTHERN DIVISION', 'MOROTO'),
 ('SOUTHERN DIVISION', 'MOROTO'),
 ('LOREGAE', 'NAKAPIRIPIRIT'),
 ('NAKAPIRIPIRIT TOWN COUNCIL', 'NAKAPIRIPIRIT'),
 ('NAMALU', 'NAKAPIRIPIRIT'),
 ('KAKOMONGOLE', 'NAKAPIRIPIRIT'),
 ('LOLACHAT', 'NAKAPIRIPIRIT'),
 ('LORENGEDWAT', 'NAKAPIRIPIRIT'),
 ('NABILATUK', 'NAKAPIRIPIRIT'),
 ('IRIIRI', 'NAPAK'),
 ('LOPEEI', 'NAPAK'),
 ('LORENGECORA', 'NAPAK'),
 ('LOTOME', 'NAPAK'),
 ('MATANY', 'NAPAK'),
 ('NAPAK TOWN COUNCIL', 'NAPAK'),
 ('NGOLERIET', 'NAPAK'),
 ('NYAKWAE', 'ABIM'),
 ('LOKOPO', 'NAPAK'),
 ('RUPA', 'MOROTO'),
 ('MORUITA', 'NAKAPIRIPIRIT')]

In [ ]:
#select all the columns, but return only one record
%sql SELECT * FROM uganda LIMIT 1;

 * sqlite://
Done.


[(263, 'KACHERI', 'KOTIDO', 17244, 1067176155, 'Y', 354.20741069999997, 1137.467019, 7023.533691, 6434.342449, 528.1242294, 2279091.779, 600723.8929)]

**<h4> IP QUESTIONS <h4>**

**<h4> 1.1 Display a list of Sub Counties and their population and areas.<h4>**

In [ ]:
#question 1

%%sql
SELECT SUBCOUNTY_NAME, POP, Area FROM uganda;

 * sqlite://
Done.


[('KACHERI', 17244, 1067176155),
 ('KOTIDO', 52771, 597575188),
 ('KOTIDO TOWN COUNCIL', 27389, 23972401),
 ('NAKAPERIMORU', 38775, 419111591),
 ('PANYANGARA', 65704, 880955930),
 ('RENGEN', 41273, 652744859),
 ('KAABONG  EAST', 42221, 60801942),
 ('KAABONG TOWN COUNCIL', 38857, 13071455),
 ('KAABONG WEST', 41454, 67612362),
 ('KALAPATA', 99203, 223116860),
 ('KAMION', 60070, 1199409465),
 ('KAPEDO', 33951, 304557831),
 ('KATHILE', 56870, 228457417),
 ('KAWALAKOL', 37691, 1742041261),
 ('LOBALANGIT', 22907, 302572789),
 ('LODIKO', 31820, 397700845),
 ('LOLELIA', 100919, 1069331387),
 ('LOYORO', 16608, 1330019236),
 ('NAPORE (KARENGA)', 20394, 309577387),
 ('SIDOK', 24092, 125330583),
 ('ABIM', 14377, 188613057),
 ('ABIM TOWN COUNCIL', 11489, 115941152),
 ('ALEREK', 12573, 1255760525),
 ('LOTUKEI', 18641, 191684465),
 ('MORULEM', 16834, 250364895),
 ('KARITA', 45015, 658008801),
 ('LOROO', 22541, 371970302),
 ('AMUDAT', 30181, 596393139),
 ('AMUDAT TOWN COUNCIL', 4053, 17207110),
 ('KATIKEKILE', 11772, 277451997),
 ('NADUNGET', 42115, 596023780),
 ('TAPAC', 21999, 622266516),
 ('NORTHERN DIVISION', 7161, 2738777),
 ('SOUTHERN DIVISION', 3272, 2121209),
 ('LOREGAE', 26644, 928683431),
 ('NAKAPIRIPIRIT TOWN COUNCIL', 1418, 4673094),
 ('NAMALU', 25905, 553067364),
 ('KAKOMONGOLE', 15960, 312709097),
 ('LOLACHAT', 22801, 642401705),
 ('LORENGEDWAT', 9929, 435714604),
 ('NABILATUK', 27534, 499776866),
 ('IRIIRI', 46979, 1030623258),
 ('LOPEEI', 18815, 790782350),
 ('LORENGECORA', 2503, 118997955),
 ('LOTOME', 17834, 125857565),
 ('MATANY', 29495, 338252511),
 ('NAPAK TOWN COUNCIL', 8010, 150906375),
 ('NGOLERIET', 20790, 158887552),
 ('NYAKWAE', 16470, 769609687),
 ('LOKOPO', 23200, 1794470536),
 ('RUPA', 41493, 2069554899),
 ('MORUITA', 16588, 839293722)]

In [ ]:
%%sql
SELECT SUBCOUNTY_NAME, POP, Area 
FROM uganda 
LIMIT 5;

 * sqlite://
Done.


[('KACHERI', 17244, 1067176155),
 ('KOTIDO', 52771, 597575188),
 ('KOTIDO TOWN COUNCIL', 27389, 23972401),
 ('NAKAPERIMORU', 38775, 419111591),
 ('PANYANGARA', 65704, 880955930)]

QUESTION: in cell 17 and 18, why is it that the column names are not listed in the output(in cell 17) and yet when i use the same code and limit the number of records(as in cell 18, the column names are given)?

**<h4> 1.2 Sort the list of districts by total crop area (descending order).<h4>**

In [ ]:
%sql SELECT * FROM uganda LIMIT 1;

 * sqlite://
Done.


[(263, 'KACHERI', 'KOTIDO', 17244, 1067176155, 'Y', 354.20741069999997, 1137.467019, 7023.533691, 6434.342449, 528.1242294, 2279091.779, 600723.8929)]

In [ ]:
#question 2

%%sql
SELECT DISTRICT_NAME, Crop_Area_Ha FROM uganda
ORDER BY Crop_Area_Ha DESC;

 * sqlite://
Done.


[('KOTIDO', 13587.990759999999),
 ('KOTIDO', 13278.52043),
 ('KOTIDO', 10398.24939),
 ('KOTIDO', 7087.823334000001),
 ('KOTIDO', 7023.533691),
 ('KAABONG', 6890.699234000001),
 ('NAPAK', 6471.047334000001),
 ('NAKAPIRIPIRIT', 5793.5076229999995),
 ('NAKAPIRIPIRIT', 5759.618906000001),
 ('NAPAK', 5473.443674),
 ('NAKAPIRIPIRIT', 5079.779088),
 ('KAABONG', 3934.432233),
 ('NAKAPIRIPIRIT', 3276.653762),
 ('NAKAPIRIPIRIT', 3264.534948),
 ('NAPAK', 3120.318045),
 ('NAPAK', 2990.847377),
 ('KAABONG', 2834.349905),
 ('KAABONG', 2700.676313),
 ('AMUDAT', 2513.27183),
 ('MOROTO', 2480.1219),
 ('KAABONG', 2365.259806),
 ('MOROTO', 2217.2907170000003),
 ('NAKAPIRIPIRIT', 2008.0681690000001),
 ('KAABONG', 1890.248613),
 ('NAPAK', 1849.966589),
 ('KOTIDO', 1656.5318550000002),
 ('AMUDAT', 1651.99842),
 ('AMUDAT', 1539.401382),
 ('KAABONG', 1494.047587),
 ('KAABONG', 1491.771813),
 ('ABIM', 1266.109868),
 ('NAPAK', 1215.227445),
 ('NAKAPIRIPIRIT', 1161.3902289999999),
 ('MOROTO', 1111.664207),
 ('NAPAK', 1031.196285),
 ('KAABONG', 1021.31699),
 ('ABIM', 999.3715888),
 ('KAABONG', 997.1610206),
 ('KAABONG', 967.3367266),
 ('KAABONG', 957.4939457999999),
 ('ABIM', 925.0541302000001),
 ('ABIM', 792.8987889),
 ('NAPAK', 792.2492731000001),
 ('ABIM', 788.4628001),
 ('ABIM', 698.1712182),
 ('KAABONG', 401.7978207),
 ('KAABONG', 175.08052659999998),
 ('MOROTO', 145.243518),
 ('AMUDAT', 60.77208552),
 ('NAKAPIRIPIRIT', 29.14576036),
 ('MOROTO', 0.322315707),
 ('MOROTO', 0.17139046100000002)]

In [ ]:
%%sql
SELECT DISTRICT_NAME FROM uganda ASC LIMIT 3;

 * sqlite://
Done.


[('KOTIDO',), ('KOTIDO',), ('KOTIDO',)]

In [ ]:
%%sql
SELECT DISTRICT_NAME FROM uganda ASC LIMIT 5

 * sqlite://
Done.


[('KOTIDO',), ('KOTIDO',), ('KOTIDO',), ('KOTIDO',), ('KOTIDO',)]

In [ ]:
%sql SELECT * FROM uganda LIMIT 2;


 * sqlite://
Done.


[(263, 'KACHERI', 'KOTIDO', 17244, 1067176155, 'Y', 354.20741069999997, 1137.467019, 7023.533691, 6434.342449, 528.1242294, 2279091.779, 600723.8929),
 (264, 'KOTIDO', 'KOTIDO', 52771, 597575188, 'Y', 367.89052319999996, 1162.9966869999998, 13587.990759999999, 12455.59264, 824.7670809, 4582294.492, 959201.3825)]

**<h4>1.3 Select only the Sub counties from the Moroto district, order them alphabetically and show their production of sorghum.<h4>**

In [ ]:
#question 3
#select all subcounties located in Moroto district
#order the sucounties alphabetically
#also select their total sorghum production

%%sql
SELECT SUBCOUNTY_NAME, DISTRICT_NAME,S_Prod_Tot	FROM uganda
WHERE DISTRICT_NAME = 'MOROTO'
ORDER BY SUBCOUNTY_NAME ASC;

 * sqlite://
Done.


[('KATIKEKILE', 'MOROTO', 9379.029413),
 ('NADUNGET', 'MOROTO', 321630.539),
 ('NORTHERN DIVISION', 'MOROTO', 17.28125956),
 ('RUPA', 'MOROTO', 227298.5184),
 ('SOUTHERN DIVISION', 'MOROTO', 19.67486547),
 ('TAPAC', 'MOROTO', 49251.66081)]

**<h4>1.4 Compute the total Maize production per District.<h4>**

In [ ]:
#question 4
#total maize production per district
#summing up the maize productions in every single district

%%sql
SELECT DISTRICT_NAME, SUM(M_Prod_Tot)
FROM uganda 
GROUP BY DISTRICT_NAME;

 * sqlite://
Done.


[('ABIM', 1922133.1297),
 ('AMUDAT', 3546469.3324990002),
 ('KAABONG', 6990361.6143000005),
 ('KOTIDO', 2010179.0731479998),
 ('MOROTO', 422116.06882464996),
 ('NAKAPIRIPIRIT', 8125156.24546),
 ('NAPAK', 5587379.38888)]

found 7 districts and the total maize production 
in each district is as shown in the console above

**<h4>1.5 Compute the number of Sub counties where Maize is produced and the total Maize production per District.<h4>**

In [ ]:
%sql SELECT * FROM uganda LIMIT 2;

 * sqlite://
Done.


[(263, 'KACHERI', 'KOTIDO', 17244, 1067176155, 'Y', 354.20741069999997, 1137.467019, 7023.533691, 6434.342449, 528.1242294, 2279091.779, 600723.8929),
 (264, 'KOTIDO', 'KOTIDO', 52771, 597575188, 'Y', 367.89052319999996, 1162.9966869999998, 13587.990759999999, 12455.59264, 824.7670809, 4582294.492, 959201.3825)]

In [ ]:
#sub counties where maize is NOT produced
%%sql
SELECT SUBCOUNTY_NAME FROM uganda
WHERE M_Area_Ha IS NULL;


 * sqlite://
Done.


[]

In [ ]:
#sub counties where maize is NOT produced
%%sql
SELECT SUBCOUNTY_NAME, DISTRICT_NAME FROM uganda
WHERE M_Area_Ha = 0;

 * sqlite://
Done.


[('SOUTHERN DIVISION', 'MOROTO')]

maize is not produced in SOUTHERN DIVISION subcounty of MOROTO district as shown in the cells(avove and below)

In [ ]:
#sub counties where maize is NOT produced
%%sql
SELECT SUBCOUNTY_NAME, DISTRICT_NAME, M_Area_Ha, M_Prod_Tot 
FROM uganda
WHERE M_Area_Ha = 0;

 * sqlite://
Done.


[('SOUTHERN DIVISION', 'MOROTO', 0.0, 0.0)]


1.  therefore, maize is not produced in southern division county(1 SUBCOUNTY)
2.   we had 52 subcounties in total in the table

3.  this means (52-1)= 51 sub counties produce maize
4.   lest check it in the code below






In [ ]:
#QUESTION 5
%%sql 
SELECT COUNT(SUBCOUNTY_NAME), DISTRICT_NAME, SUM(M_Prod_Tot)
FROM uganda 
WHERE M_Area_Ha != 0
GROUP BY DISTRICT_NAME;

 * sqlite://
Done.


[(6, 'ABIM', 1922133.1297),
 (4, 'AMUDAT', 3546469.3324990002),
 (14, 'KAABONG', 6990361.6143000005),
 (6, 'KOTIDO', 2010179.0731479998),
 (5, 'MOROTO', 422116.06882464996),
 (8, 'NAKAPIRIPIRIT', 8125156.24546),
 (8, 'NAPAK', 5587379.38888)]

summing up the subcounties above, (6+4+14+6+5+8+8)=51 subcounties producing maize

this means, for example,there are 6 subcounties in ABIM district and the total maize produce in that district is 1922133.1297

**<h4>1.6 Compute the overall Crop area in all Sub counties where population is over 20000.<h4>**

In [ ]:
#question 6
#crop area in subcounties where population>20000

%%sql
SELECT SUBCOUNTY_NAME, Crop_Area_Ha, POP FROM uganda
WHERE POP > 20000;



 * sqlite://
Done.


[('KOTIDO', 13587.990759999999, 52771),
 ('KOTIDO TOWN COUNCIL', 1656.5318550000002, 27389),
 ('NAKAPERIMORU', 7087.823334000001, 38775),
 ('PANYANGARA', 10398.24939, 65704),
 ('RENGEN', 13278.52043, 41273),
 ('KAABONG  EAST', 1021.31699, 42221),
 ('KAABONG TOWN COUNCIL', 175.08052659999998, 38857),
 ('KAABONG WEST', 967.3367266, 41454),
 ('KALAPATA', 2834.349905, 99203),
 ('KAMION', 997.1610206, 60070),
 ('KAPEDO', 2365.259806, 33951),
 ('KATHILE', 6890.699234000001, 56870),
 ('KAWALAKOL', 3934.432233, 37691),
 ('LOBALANGIT', 1890.248613, 22907),
 ('LODIKO', 401.7978207, 31820),
 ('LOLELIA', 2700.676313, 100919),
 ('NAPORE (KARENGA)', 1494.047587, 20394),
 ('SIDOK', 957.4939457999999, 24092),
 ('KARITA', 2513.27183, 45015),
 ('LOROO', 1539.401382, 22541),
 ('AMUDAT', 1651.99842, 30181),
 ('NADUNGET', 2480.1219, 42115),
 ('TAPAC', 1111.664207, 21999),
 ('LOREGAE', 5759.618906000001, 26644),
 ('NAMALU', 5793.5076229999995, 25905),
 ('LOLACHAT', 5079.779088, 22801),
 ('NABILATUK', 3276.653762, 27534),
 ('IRIIRI', 5473.443674, 46979),
 ('MATANY', 2990.847377, 29495),
 ('NGOLERIET', 1215.227445, 20790),
 ('LOKOPO', 6471.047334000001, 23200),
 ('RUPA', 2217.2907170000003, 41493)]

**<h4>1.7 Sort the Maize production in descending order by Districts, only taking into account Sub counties where Maize area is larger than Sorghum area, and display the number of Sub counties per district matching that criteria.<h4>**

In [ ]:
%sql SELECT * FROM uganda LIMIT 2;


 * sqlite://
Done.


[(263, 'KACHERI', 'KOTIDO', 17244, 1067176155, 'Y', 354.20741069999997, 1137.467019, 7023.533691, 6434.342449, 528.1242294, 2279091.779, 600723.8929),
 (264, 'KOTIDO', 'KOTIDO', 52771, 597575188, 'Y', 367.89052319999996, 1162.9966869999998, 13587.990759999999, 12455.59264, 824.7670809, 4582294.492, 959201.3825)]

In [ ]:
%%sql
SELECT SUBCOUNTY_NAME, DISTRICT_NAME FROM uganda
WHERE M_Area_Ha > S_Area_Ha
ORDER BY DISTRICT_NAME DESC LIMIT 5;




 * sqlite://
Done.


[('IRIIRI', 'NAPAK'),
 ('NAPAK TOWN COUNCIL', 'NAPAK'),
 ('NAKAPIRIPIRIT TOWN COUNCIL', 'NAKAPIRIPIRIT'),
 ('NAMALU', 'NAKAPIRIPIRIT'),
 ('MORUITA', 'NAKAPIRIPIRIT')]

In [ ]:
#question 7
%%sql
SELECT COUNT(SUBCOUNTY_NAME), DISTRICT_NAME,Crop_Area_Ha, S_Area_Ha, M_Area_Ha FROM uganda
WHERE M_Area_Ha > S_Area_Ha
GROUP BY DISTRICT_NAME
ORDER BY DISTRICT_NAME DESC;


 * sqlite://
Done.


[(2, 'NAPAK', 792.2492731000001, 386.1580327, 401.1734419),
 (3, 'NAKAPIRIPIRIT', 1161.3902289999999, 185.2834449, 959.6711616),
 (3, 'MOROTO', 0.322315707, 0.130940756, 0.19137495100000002),
 (3, 'KAABONG', 1494.047587, 267.559128, 1221.532726),
 (1, 'AMUDAT', 1651.99842, 676.7832664, 972.1742458999998)]

This means five districts satisfy the criteria(M_Area_Ha > S_Area_Ha)
The number of subcounties per districts that satisfy the criteria is as displayed in the console above e.g 2 sub counties in NAPAK district satisfy the criteria

